Change log level of error messages logged on execution failure from ERROR to DEBUG

dmoidl opened the pull request:
A solution for [this issue](https://github.com/kagkarlsson/db-scheduler/issues/170).

dmoidl committed:
Change log level of error messages logged on execution failure from ERROR to DEBUG

dmoidl commented on db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/Scheduler.java:
```diff
@@ -356,12 +356,12 @@ private void executePickedExecution(Execution execution) {
 
             } catch (RuntimeException unhandledException) {
-                LOG.error("Unhandled exception during execution of task with name '{}'. Treating as failure.", task.get().getName(), unhandledException);
+                LOG.debug("Unhandled exception during execution of task with name '{}'. Treating as failure.", task.get().getName(), unhandledException);
```
I can imagine also changing the log level in the default `FailureHandler` implementations (i.e. `OnFailureRetryLater` and `OnFailureReschedule`) to `ERROR` to keep pretty much the same experience for simple tasks using those built-ins, but it's not directly related to the problem this PR aims to solve so I didn't.

dmoidl replied:
We discussed this and decided not to go this way.

dmoidl committed:
Make failure logging configurable

dmoidl commented on db-scheduler/src/test/java/com/github/kagkarlsson/scheduler/logging/LogHelperTest.java:
```diff
@@ -0,0 +1,46 @@
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.Is.is;
+
+public class LogHelperTest {
```
I hope a test of this kind is enough. I didn't really want to refactor much of the existing test code...

kagkarlsson replied:
Yeah, that should be fine. No need to verify in larger tests I think.

dmoidl reviewed (commented):
Here's the implementation for configurable failure logging as I promised 😉

kagkarlsson commented on db-scheduler/pom.xml:
```diff
@@ -22,7 +22,7 @@
         <otj-pg-embedded.version>0.11.4</otj-pg-embedded.version>
         <postgresql.version>9.4.1207</postgresql.version>
-        <slf4j.version>1.7.7</slf4j.version>
+        <slf4j.version>1.7.30</slf4j.version>
```
Will that work for users specifying an older version in their pom?

kagkarlsson replied:
On second thought I think it is better to not have a dependency on slf4f in the Builder-interface. I think we should create our own enum :)

dmoidl replied:
Done, new enum created.

kagkarlsson commented on db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/SchedulerBuilder.java:
```diff
@@ -163,6 +167,15 @@ public SchedulerBuilder commitWhenAutocommitDisabled(boolean commitWhenAutocommi
 
+    public SchedulerBuilder failureLogging(Level logLevel, boolean logStackTrace) {
+        if(logLevel == null) {
+            throw new IllegalArgumentException("Log level must not be null");
```
Looks good the way it is 👍

kagkarlsson commented on db-scheduler-boot-starter/src/main/java/com/github/kagkarlsson/scheduler/boot/config/DbSchedulerProperties.java:
```diff
@@ -106,6 +107,17 @@
+     * <p>Which log level to use when logging task failures. Defaults to {@link Level#DEBUG}.</p>
+     */
+    @NotNull
+    private Level failureLogLevel = Level.DEBUG;
+
+    /**
+     * <p>Whether or not to log the {@link Throwable} that caused a task to fail.</p>
+     */
+    private boolean logStackTraceOnFailure = false;
+
```
These need to refer to the constants in the `SchedulerBuilder` to ensure we get the same default

kagkarlsson commented on db-scheduler-boot-starter/src/main/java/com/github/kagkarlsson/scheduler/boot/config/DbSchedulerProperties.java:
```diff
@@ -20,6 +20,7 @@
 import java.time.Duration;
 import java.util.Optional;
 import javax.validation.constraints.NotNull;
+import org.slf4j.event.Level;
```
As I mentioned, I think we need to create our own enum. It's cleaner :)

kagkarlsson commented on db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/Scheduler.java:
```diff
@@ -60,9 +62,12 @@
     private final SettableSchedulerState schedulerState = new SettableSchedulerState();
     private int currentGenerationNumber = 1;
+    private final LogHelper.LogMethod failureLogMethod;
+    private final boolean logStackTrace;
 
     protected Scheduler(Clock clock, TaskRepository schedulerTaskRepository, TaskRepository clientTaskRepository, TaskResolver taskResolver, int threadpoolSize, ExecutorService executorService, SchedulerName schedulerName,
-                        Waiter executeDueWaiter, Duration heartbeatInterval, boolean enableImmediateExecution, StatsRegistry statsRegistry, int pollingLimit, Duration deleteUnresolvedAfter, Duration shutdownMaxWait, List<OnStartup> onStartup) {
+                        Waiter executeDueWaiter, Duration heartbeatInterval, boolean enableImmediateExecution, StatsRegistry statsRegistry, int pollingLimit, Duration deleteUnresolvedAfter, Duration shutdownMaxWait, List<OnStartup> onStartup,
+                        Level logLevel, boolean logStackTrace) {
         this.clock = clock;
```
At some point I need to refactor this constructor 😊

kagkarlsson commented on db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/Scheduler.java:
```diff
@@ -82,6 +87,8 @@ protected Scheduler(Clock clock, TaskRepository schedulerTaskRepository, TaskRep
         this.updateHeartbeatExecutor = Executors.newSingleThreadExecutor(defaultThreadFactoryWithPrefix(THREAD_PREFIX + "-update-heartbeat-"));
         SchedulerClientEventListener earlyExecutionListener = (enableImmediateExecution ? new TriggerCheckForDueExecutions(schedulerState, clock, executeDueWaiter) : SchedulerClientEventListener.NOOP);
         delegate = new StandardSchedulerClient(clientTaskRepository, earlyExecutionListener);
+        this.failureLogMethod = LogHelper.getLogMethod(LOG, logLevel);
+        this.logStackTrace = logStackTrace;
```
We might make a bit more pretty if we introduce a wrapper-object, something like

```
failureLogger = new FailureLogger(LOG, logLevel, logStackTrace);
```
...
```
class FailureLogger
  logFailure(String message, Throwable, Object .. messageContext)
  ```

dmoidl replied:
Implemented.

kagkarlsson commented on db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/SchedulerBuilder.java:
```diff
@@ -190,7 +203,7 @@ public Scheduler build() {
-                schedulerName, waiter, heartbeatInterval, enableImmediateExecution, statsRegistry, pollingLimit,
-            deleteUnresolvedAfter, shutdownMaxWait, startTasks);
+            schedulerName, waiter, heartbeatInterval, enableImmediateExecution, statsRegistry, pollingLimit,
+            deleteUnresolvedAfter, shutdownMaxWait, startTasks, logLevel, logStackTrace);
```
I think we should let `startTasks` be the last parameter here

kagkarlsson commented on db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/logging/LogHelper.java:
```diff
@@ -0,0 +1,50 @@
+        void log(String format, Object... arguments);
+    }
+
+    public static LogMethod getLogMethod(Logger logger, Level logLevel) {
```
Yeah, nice use of method references 👍 , but see my suggestion on a `FailureLogger` class. (might still get to use some of this code though)

dmoidl replied:
Implemented the `FailureLogger` class as per your suggestion 🙂

kagkarlsson reviewed (commented):
Looks good, but a couple of comments :) 

Btw, heads-up, I have some changes coming up as well. Depending on timing here, it might generate some small merge-work for this branch

dmoidl committed:
Add FailureLogger

dmoidl commented:
Sorry for force-pushing, but I had some conflicts with `master` and I prefer rebasing to keep the history clean 🙂

dmoidl commented on db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/FailureLogger.java:
```diff
@@ -0,0 +1,71 @@
+import com.github.kagkarlsson.scheduler.logging.LogLevel;
+import org.slf4j.Logger;
+
+class FailureLogger {
```
I'd like to keep this class in the `logging` package, but at the same time I wanted it to be package-private as this is (arguably) not something we want to expose to the users... Or we could move the `LogLevel` to the main package and remove the `logging` package altogether. WDYT? 🤔

kagkarlsson replied:
It doesn't really bother me if it would be public, I haven't really been strict about that in other places. Also, it turned out quite generic, so `.logging` sounds good

kagkarlsson replied:
It became something more like `ConfigurableLogger` rather than something having anything to do with Failure :)

dmoidl replied:
`ConfigurableLogger` it is, then.

kagkarlsson commented on db-scheduler-boot-starter/src/main/java/com/github/kagkarlsson/scheduler/boot/config/DbSchedulerProperties.java:
```diff
@@ -98,6 +100,16 @@
+    /**
+     * <p>Which log level to use when logging task failures. Defaults to {@link LogLevel#DEBUG}.</p>
+     */
+    private LogLevel failureLogLevel = SchedulerBuilder.DEFAULT_FAILURE_LOG_LEVEL;
+
+    /**
+     * <p>Whether or not to log the {@link Throwable} that caused a task to fail.</p>
+     */
+    private boolean logStackTraceOnFailure = SchedulerBuilder.LOG_STACK_TRACE_ON_FAILURE;
+
```
I was thinking if maybe these two properties should have the same prefix, so you can see they belong together in the spring properties file. Like `failureLoggerLevel`, `failureLoggerLogStacktrace` ?

dmoidl replied:
I was thinking the same but couldn't come up with a good name for the latter property 🙃  The ones that you proposed sound good though, code updated.

dmoidl committed:
Minor refactoring

dmoidl commented:
Looks like the failing checks are not caused by my changes...

kagkarlsson commented:
Upgrading to testcontainers 1.15.1, that should fix it.

kagkarlsson commented:
Having some seconds thoughts on what should be the default log-level. Would not WARN be a safer default, and then users' who are bothered by it can tune it explicitly to DEBUG? I worry that some users' will miss important exceptions using the default-config (i.e. it is not "safe")

dmoidl committed:
Add option to turn failure logging off

dmoidl committed:
Change default failure log level to WARN

dmoidl commented:
> Having some seconds thoughts on what should be the default log-level. Would not WARN be a safer default, and then users' who are bothered by it can tune it explicitly to DEBUG? I worry that some users' will miss important exceptions using the default-config (i.e. it is not "safe")

Sure, I think this all boils down to how you'd like the people to use the library 🙂  In my eyes, it really doesn't matter all that much since now everyone can easily adjust the logging as they see fit, but you're right that `DEBUG` may be too low if you want to provide some information to the user using the out-of-the-box settings.

I've changed the default to `WARN` and also added a new log level `OFF` to turn the logging off completely so that it's 100% customizable 🙂

kagkarlsson commented:
Excellent! I will merge this very soon 👍

dmoidl committed:
Fix ConfigurableLoggerTest

dmoidl commented:
@kagkarlsson Sorry, didn't notice those failing tests. Fix pushed 🙂

kagkarlsson commented:
🎉 This issue has been resolved in `v16.9.0` ([Release Notes](https://github.com/kagkarlsson/db-scheduler/releases/tag/v16.9.0))
